f/fcs/fn_fcsInit.sqf

```diff
@@ -24,83 +24,93 @@
-		[_target,[false,[0]]] remoteExec ["enableDirectionStabilization",0,_target];
-		_target disableNVGEquipment false;
-	}, // Code on completed
-	{}, // Code on interrupt
-	[], // Arguments to pass
-	10, // Duration
-	1, // Priority
-	false, // Remove on completion
-	false, // Show when unconscious
-	true // Show on screen
-] call BIS_fnc_holdActionAdd;
+	// Hit EH for FCS failure
+	_vehicle addEventHandler ["HitPart",{
+			(_this select 0) params ["_target", "_shooter", "_projectile", "_position", "_velocity", "_selection", "_ammo", "_vector", "_radius", "_surfaceType", "_isDirect"];
+			// Call the actual failure evaluator
+			[_target,_projectile,_isDirect] call f_fnc_fcsFailure;
+		}];
+		
+	_vehicle setVariable ["fcs_hasEH",true];
 
-// Action to prevent rangefinder use during FCS failure
-_vehicle addAction ["FCS failure - repair required", // Title
-	{
-		hint "Someone familiar with maintenance can reset the FCS.";
-	}, // Code
-	"", // Arguments
-	10, // Priority
-	false, // Show window
-	true, // Hide on use
-	"gunElevAuto", // Shortcut
-	"(_this == gunner _target) && {_target getVariable [""fcs_failure"",false]}" // Condition
-];
+	// Hold action to repair FCS failure
+	// Players with repair trait in the gunner's seat of a vehicle with failed stabiliser can repair it
+	[
+		_vehicle, // Target
+		"Reset Fire Control System", // Title
+		"\a3\ui_f_oldman\data\IGUI\Cfg\holdactions\repair_ca.paa", // Idle icon
+		"\a3\ui_f_oldman\data\IGUI\Cfg\holdactions\repair_ca.paa", // Progress icon
+		"(_target getVariable [""f3_fcs_failure"",false]) && (gunner _target == _this) && (_this getUnitTrait ""engineer"")", // Condition to show
+		"(_target getVariable [""f3_fcs_failure"",false]) && (gunner _target == _this) && (_this getUnitTrait ""engineer"")", // Condition to progress
+		{}, // Code on start
+		{}, // Code on tick
+		{ 
+			params ["_target", "_caller", "_actionId", "_arguments"];
+			_target setVariable ["f3_fcs_failure",false,true];
+			[_target,[false,[0]]] remoteExec ["enableDirectionStabilization",0,_target];
+			_target disableNVGEquipment false;
+		}, // Code on completed
+		{}, // Code on interrupt
+		[], // Arguments to pass
+		10, // Duration
+		1, // Priority
+		false, // Remove on completion
+		false, // Show when unconscious
+		true // Show on screen
+	] call BIS_fnc_holdActionAdd;
+
+	// Action to prevent rangefinder use during FCS failure
+	_vehicle addAction ["FCS failure - repair required", // Title
+		{
+			hint "Someone familiar with maintenance can reset the FCS.";
+		}, // Code
+		"", // Arguments
+		10, // Priority
+		false, // Show window
+		true, // Hide on use
+		"gunElevAuto", // Shortcut
+		"(_this == gunner _target) && {_target getVariable [""f3_fcs_failure"",false]}" // Condition
+	];
+};
+
+if (_vehicle getVariable ["f3_fcs_hasDriverAction",false]) exitWith { diag_log "FCS: tried to add driver action on something that already has it"};
 
 // Not FCS but I'm putting it in here
 _vehicle addAction [
```

Comment:
Question: continuation of the idea of this being a separate component, is there a reason not to add this to every (land) vehicle? With that the action could be added to the player instead and just act on any vehicle they are the driver of.

Reply:
It is possible, although unarmed vehicles don't really benefit from turning off the autobrakes, which is why I didn't bother with it. While it's not "FCS" in the technical sense, it's part of a suite of improvements targeted at armed vehicles in particular - for example, the advanced FCS stuff is useless for technicals but the brake override is useful. So this got added from the side of "expand the gunnery improvements" rather than as a separate thing on its own.

So I guess it comes down to whether to consider it a utility for armed vehicles or whether it's actually a separate thing all vehicles should have.